### ISSUE TITLE
Using Aruba in directories with spaces

### ISSUE THREAD

**Issue by odlp** [cucumber/aruba#490]
## Summary

Aruba (`0.14.2`) isn't able to spawn processes within folders which have spaces in the name.

## Expected Behavior

```sh
$ pwd
/Users/oli/workspace/aruba-spaces-bug/regrettable folder with spaces

$ bundle exec cucumber -b

Feature: Example

  Scenario: Running in a directory with spaces # features/example.feature:2
    When I successfully run `cli`              # aruba-0.14.2/lib/aruba/cucumber/command.rb:27
    Then the output should contain:            # aruba-0.14.2/lib/aruba/cucumber/command.rb:205
      """
      foo
      """

1 scenario (1 passed)
2 steps (2 passed)
0m0.113s
```

## Current Behavior

```sh
$ pwd
/Users/oli/workspace/aruba-spaces-bug/regrettable folder with spaces

$ bundle exec cucumber -b

Feature: Example

  Scenario: Running in a directory with spaces # features/example.feature:2
    When I successfully run `cli`              # aruba-0.14.2/lib/aruba/cucumber/command.rb:27
      undefined local variable or method `cmd' for #<Aruba::Processes::SpawnProcess:0x007f95102b2328>
      Did you mean?  @cmd (NameError)
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/processes/spawn_process.rb:86:in `rescue in start'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/processes/spawn_process.rb:80:in `start'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/command.rb:67:in `start'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/api/command.rb:213:in `run'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/api/command.rb:271:in `run_simple'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/cucumber/command.rb:29:in `block in <top (required)>'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/cucumber-2.4.0/lib/cucumber/core_ext/instance_exec.rb:25:in `instance_exec'
```

## Possible Solution

First-off the exception is obscured due to an undefined variable / method `cmd`:

https://github.com/cucumber/aruba/blob/70a78abfaaed2bafa16b6ebf21ac9b56d95f8534/lib/aruba/processes/spawn_process.rb#L86

However this has been corrected to `commandline` in subsequent releases of Aruba. Correcting this line locally, and re-running `bundle exec cucumber -b`:

```
Feature: Example

  Scenario: Running in a directory with spaces # features/example.feature:2
    When I successfully run `cli`              # aruba-0.14.2/lib/aruba/cucumber/command.rb:27
      It tried to start cli. No such file or directory - /Users/oli/workspace/aruba-spaces-bug/regrettable (Aruba::LaunchError)
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/processes/spawn_process.rb:86:in `rescue in start'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/processes/spawn_process.rb:80:in `start'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/command.rb:67:in `start'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/api/command.rb:213:in `run'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/api/command.rb:271:in `run_simple'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/aruba-0.14.2/lib/aruba/cucumber/command.rb:29:in `block in <top (required)>'
```

The notable part here is the directory is incorrect:

```
No such file or directory - /Users/oli/workspace/aruba-spaces-bug/regrettable
```

It should be:
```
/Users/oli/workspace/aruba-spaces-bug/regrettable\ folder\ with\ spaces
```

This could be because the `@process.cwd` needs to be escaped with [`Shellwords.escape`](http://ruby-doc.org/stdlib-2.4.1/libdoc/shellwords/rdoc/Shellwords.html#method-c-shellescape) or similar:
https://github.com/cucumber/aruba/blob/70a78abfaaed2bafa16b6ebf21ac9b56d95f8534/lib/aruba/processes/spawn_process.rb#L76

And possibly the `command_string` before being passed to `ChildProcess` too:

https://github.com/cucumber/aruba/blob/70a78abfaaed2bafa16b6ebf21ac9b56d95f8534/lib/aruba/processes/spawn_process.rb#L60

## Steps to Reproduce (for bugs)

```
git clone https://github.com/odlp/aruba-spaces-bug.git
cd aruba-spaces-bug
bundle install

# Works
cd folder-without-spaces
bundle exec cucumber

# Fails
cd ../regrettable\ folder\ with\ spaces/
bundle exec cucumber
```

## Context & Motivation

Ideally users wouldn't be in directories with spaces, but this can happen when the same project is unzipped multiple times for example. OSX in particular creates successive filenames like: `your_project`, `your_project (1)`, `your_project (2)`.

**Comment by olleolleolle** [comment]
Could you test the same thing, with a `master` aruba?

That could be done changing this line:
https://github.com/odlp/aruba-spaces-bug/blob/master/Gemfile#L4 to point to master

Link: [Gems from git repositories](http://bundler.io/v1.14/git.html)

**Comment by odlp** [comment]
@olleolleolle I've just tried with `master @ ` https://github.com/cucumber/aruba/commit/6820ab32e9a855ba5be1495e516a3075e0cb178d and still get an error:

```
Feature: Example

  Scenario: Running in a directory with spaces # features/example.feature:2
    When I successfully run `cli`              # aruba-6820ab32e9a8/lib/aruba/cucumber/command.rb:24
      It tried to start cli. No such file or directory - /Users/oli/workspace/aruba-spaces-bug/regrettable (Aruba::LaunchError)
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/bundler/gems/aruba-6820ab32e9a8/lib/aruba/processes/spawn_process.rb:89:in `rescue in start'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/bundler/gems/aruba-6820ab32e9a8/lib/aruba/processes/spawn_process.rb:83:in `start'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/bundler/gems/aruba-6820ab32e9a8/lib/aruba/command.rb:67:in `start'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/bundler/gems/aruba-6820ab32e9a8/lib/aruba/api/command.rb:215:in `run_command'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/bundler/gems/aruba-6820ab32e9a8/lib/aruba/api/command.rb:273:in `run_command_and_stop'
      /Users/oli/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/bundler/gems/aruba-6820ab32e9a8/lib/aruba/cucumber/command.rb:26:in `block in <top (required)>'
```

Looks like the process is being spawned in a folder which doesn't exist because the path is truncated from the first space character.

**Comment by olleolleolle** [comment]
@odlp Thanks, now you've verified the issue still exists on `master`. That's super - the stacktrace is also valuable.

**Comment by maxmeyer** [comment]
@odlp Would you love to help out and try to fix this bug yourself?

**Comment by mvz** [comment]
This may not be so easy: This may be a problem in the `childprocess` gem.

**Comment by maxmeyer** [comment]
And this seems to be more or less unmaintained as @enkessler doesn't have the resources to work on it: https://cucumberbdd.slack.com/archives/C62GZFLLT/p1509455288000065

**Comment by mvz** [comment]
I've been investigating if childprocess is still needed, since the built-in tools for Ruby have improved over the years. No conclusion yet, though.

**Comment by enkessler** [comment]
All right, I'll bite. What leads you to suspect that the problem is down a level in `childprocess`? Is the command being executed legit on its own? Can you `system` the command and get correct results?

**Comment by mvz** [comment]
@enkessler It's mainly a hunch from what I learned about its inner workings while trying to get the Aruba Windows build on appveyor to run, in particular how it quotes the command and arguments. However, I haven't looked at how childprocess works on Linux.

> Can you system the command and get correct results?

Well, system works with commands that contain spaces, as long as I trick it into not starting a subshell. I'll see if I can test the exact command that fails in the tests.

**Comment by mvz** [comment]
> I'll see if I can test the exact command that fails in the tests.

Well, that was very insightful. I'm preparing a fix now.

**Comment by mvz** [comment]
@enkessler it turns out this particular issue is not due to childprocess. The problem on Windows is a different one. Since I don't run Windows locally breaking this down to a concrete bug report for childprocess is a bit tricky.

**Comment by mvz** [comment]
Fixed in master. Needs fix on `0.14.x`

**Comment by mvz** [comment]
Fixed in still. Will be released in 0.14.4.
